jxl/src/api/inner/codestream_parser.rs

```diff
@@ -193,21 +212,84 @@ impl CodestreamParser {
                 // Trying to read a frame or a file header.
                 assert!(self.frame.is_none());
                 assert!(self.has_more_frames);
-                let available_codestream = box_parser.get_more_codestream(input)? as usize;
-                let c = self.non_section_buf.refill(
-                    |buf| {
-                        if !box_parser.box_buffer.is_empty() {
-                            let c = box_parser.box_buffer.take(buf);
-                            Ok(box_parser.box_buffer.consume(c))
+
```

Comment:
Can you explain why this change is necessary? I find suspicious that it would be, or at least that such a large change would be necessary.

Reply:
OK, the change is smaller now. WDYT?

Reply:
If non_section_buf contains a header, all the sections, and the beginnings of the next header, then process_sections will only move the section data from non_section_buf, and then another frame header will be read, which we have to look for first in the non_section_buf before we refill it from the box parser.

Reply:
Ok, I understand the change now, but I am not sure I like it :-)

Doesn't it work to just change line 247 to not return if Err::OutOfBounds, but instead saying that `available_codestream` is then `0`?

Reply:
OK, look again - I think it's getting pretty good.